<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <div id="root"></div>
    <script src="https://unpkg.com/react@18.1.0/umd/react.development.js"></script>
    <script src="https://unpkg.com/react-dom@18.1.0/umd/react-dom.development.js"></script>
    <script src="https://unpkg.com/@babel/standalone@7.12.4/babel.js"></script>
    <script type="text/babel">

        class ErrorBoundary extends React.Component {
            state = { error: null }
            static getDerivedStateFromError(error) {
                return { error }
            }
            render() {
                const { error } = this.state
                if (error) {
                    return <this.props.FallbackComponent error={error} />
                }

                return this.props.children
            }
        }
        const divStyle = {
            display: 'flex',
            alignItems: 'center',
            flexDirection: 'column',
            justifyContent: 'center',
            marginTop: '5rem'
        };
        const containerStyle = {
            display: 'flex',
            alignItems: 'center',
            flexDirection: 'column',
            justifyContent: 'center',
            height: '100vh'
        };

        function BuggyCounter() {
            const [counter, setCounter] = React.useState(0)
            function handleChange(event) {                
                setCounter(counter + 1)
            }
            if (counter === 6) {
                throw new Error('I crashed!');
            }
            return (
                <div style={divStyle} onClick={handleChange}>
                    This is counter but has it limits you could check the limit by clicking on it: {counter}.
                </div>
            )
        }

        function ErrorFallback({ error }) {
            return (
                <div role="alert">
                    There was an error:{' '}
                    <pre style={{ whiteSpace: 'normal' }}>{error.message}</pre>
                </div>
            )
        }

        function App() {
            return (
                <div style={containerStyle}>
                    <ErrorBoundary FallbackComponent={ErrorFallback}>
                        <BuggyCounter />
                    </ErrorBoundary>
                </div>
            )
        }
        const root = ReactDOM.createRoot(document.getElementById("root"));
        root.render(
            <React.StrictMode>
                <App />
            </React.StrictMode>
        );
    </script>
</body>

</html>